src/Libraries/Microsoft.Extensions.AI.OpenAI/OpenAIModelMapper.ChatMessage.cs

```diff
@@ -6,13 +6,97 @@
 
 using System;
 using System.Collections.Generic;
+using System.Linq;
 using System.Text.Json;
 using OpenAI.Chat;
 
 namespace Microsoft.Extensions.AI;
 
 internal static partial class OpenAIModelMappers
 {
+    internal static IEnumerable<ChatMessage> FromOpenAIChatMessages(IEnumerable<OpenAI.Chat.ChatMessage> inputs, JsonSerializerOptions options)
+    {
+        // Maps all of the OpenAI types to the corresponding M.E.AI types.
+        // Unrecognized or non-processable content is ignored.
+
+        foreach (OpenAI.Chat.ChatMessage input in inputs)
+        {
+            switch (input)
+            {
+                case SystemChatMessage systemMessage:
+                    yield return new ChatMessage
+                    {
+                        Role = ChatRole.System,
+                        AuthorName = systemMessage.ParticipantName,
+                        Contents = FromOpenAIChatContent(systemMessage.Content),
+                    };
+                    break;
+
+                case UserChatMessage userMessage:
+                    yield return new ChatMessage
+                    {
+                        Role = ChatRole.User,
+                        AuthorName = userMessage.ParticipantName,
+                        Contents = FromOpenAIChatContent(userMessage.Content),
+                    };
+                    break;
+
+                case ToolChatMessage toolMessage:
+                    string textContent = string.Join(string.Empty, toolMessage.Content.Where(part => part.Kind is ChatMessageContentPartKind.Text).Select(part => part.Text));
+                    object? result = textContent;
+                    if (!string.IsNullOrEmpty(textContent))
+                    {
+#pragma warning disable CA1031 // Do not catch general exception types
+                        try
+                        {
+                            result = JsonSerializer.Deserialize(textContent, options.GetTypeInfo(typeof(object)));
```

Comment:
I'm not certain what benefit we provide by doing this deserialization to an unknown type. In practice it will always come out as a `JsonElement` or something, right? If so that means anyone processing the `FunctionResultContent` will still have to deserialize to a particular type to do anything with it, so it makes me wonder why not just leave it as a string unconditionally.

Perhaps the benefit of deserializing to `JsonElement` is that anyone who re-serializes the `FunctionResultContent` will get a correctly round-tripped object instead of unintentionally converting the incoming object into a nested JSON string. Is that the reasoning?

Reply:
This is just performing the inverse operation of the existing `ToOpenAIChatMessages`. I think `JsonElement` is appropriate because it preserves the invariant of the `FunctionResultContent`s returned by `AIFunctionFactory` and the invoking middleware